File: examples/go/checkpoint_consumer.go
package main

import (
	"fmt"
	"os"
	"time"

	ce "auctra_examples/auctra"
)

const checkpointPath = "/tmp/orders-consumer.ckpt"

func loadOrCapture(engine *ce.Engine, path string) (*ce.Checkpoint, error) {
	cp, err := ce.LoadCheckpoint(path)
	if err == nil {
		return cp, nil
	}
	return engine.CaptureCheckpoint(true)
}

func main() {
	engine, err := ce.Open(4, 16)
	if err != nil {
		panic(err)
	}
	defer engine.Close()

	cp, err := loadOrCapture(engine, checkpointPath)
	if err != nil {
		panic(err)
	}

	for {
		items, nextCP, err := engine.ReadFromCheckpointEx(cp, 128)
		if err != nil {
			panic(err)
		}

		if len(items) == 0 {
			time.Sleep(200 * time.Millisecond)
			continue
		}

		for _, item := range items {
			fmt.Printf("seq=%d key=%s tombstone=%v\n", item.Seqno, string(item.Key), item.Tombstone)
		}

		tmp := checkpointPath + ".tmp"
		if err := ce.SaveCheckpointData(tmp, nextCP); err != nil {
			panic(err)
		}
		if err := os.Rename(tmp, checkpointPath); err != nil {
			panic(err)
		}

		cp = nextCP
	}
}

File: examples/python/checkpoint_consumer.py
import os
import time

from auctra_core import CommitEngine

CHECKPOINT_PATH = "/tmp/orders-consumer.ckpt"

def load_or_capture(engine: CommitEngine):
    if os.path.exists(CHECKPOINT_PATH):
        return engine.load_checkpoint(CHECKPOINT_PATH)
    return engine.capture_checkpoint(True)

def main():
    engine = CommitEngine(shard_count=4, inline_max=16)
    cp = load_or_capture(engine)

    while True:
        items, next_cp = engine.read_from_checkpoint(cp, limit=128)

        if not items:
            time.sleep(0.2)
            continue

        for item in items:
            print(item.seqno, item.key, item.value)

        tmp = CHECKPOINT_PATH + ".tmp"
        engine.save_checkpoint_data(tmp, next_cp)
        os.replace(tmp, CHECKPOINT_PATH)

        cp = next_cp

if __name__ == "__main__":
    main()